File: core/src/main/scala/diplomacy/variant/standard/rule/mutable/MovementOrderGroup.scala
package diplomacy.variant.standard.rule.mutable

import scala.collection.mutable
import scala.util.control.Breaks

import diplomacy.board.{ Power, DiplomacyMap }
import diplomacy.variant.standard.rule.{ Order, MilitaryBranch, Result }

final class MovementOrderGroup[Power_ <: Power](
  val target: MovementOrderWithResult[Power_],
  val relatedOrders: Set[MovementOrderWithResult[Power_]]
) extends DiplomacyMap.TypeHelper {
  type Power = Power_
  type MilitaryBranch = MilitaryBranch.MilitaryBranch

  def validSupports: Set[Order.Support[Power]] = {
    this.relatedOrders withFilter { order =>
      order.order match {
        case Order.Support(_, _) =>
          (order.result != Some(Result.Dislodged)) &&
          (order.result != Some(Result.Cut)) &&
          (order.result != Some(Result.NoCorrespondingOrder))
        case _ => false
      }
    } map { _.order } collect { case s@ Order.Support(_, _) => s }
  }

  def power: Int = 1 + this.validSupports.size

  case class Route(useConvoy: Boolean)
  def route(map: DiplomacyMap): Option[Route] = {
    this.target.order match {
      case t@Order.Move(_, _, _) =>
        def canMoveViaConvoy(
          unit: DiplomacyUnit, destination: Province, provinces: Set[Province]
        ): Boolean = {
          val provinceBuffer = mutable.Set() ++ provinces
          def dfs(current: Province): Boolean = {
            val b = new Breaks()
            var retval: Boolean = false
            b.breakable {
              map.movableProvincesOf(current, MilitaryBranch.Fleet) foreach { next =>
                if (next == destination) {
                  retval = true
                  b.break()
                } else if (provinceBuffer contains next) {
                  provinceBuffer -= next
                  if (dfs(next)) {
                    retval = true
                    b.break()
                  }
                }
              }
            }
            retval
          }

          val b = new Breaks()
          var retval: Boolean = false
          b.breakable {
            map.movableProvincesOf(unit.location.province, MilitaryBranch.Fleet) foreach { next =>
              if (provinceBuffer contains next) {
                provinceBuffer -= next
                if (dfs(next)) {
                  retval = true
                  b.break()
                }
              }
            }
          }
          retval
        }

        val provinces =
            this.relatedOrders withFilter { order =>
              order.order match {
                case Order.Convoy(_, _) =>
                  order.result match {
                    case Some(Result.Failed) | Some(Result.Success) => true
                    case _ => false
                  }
                case _ => false
              }
            } map { _.order.unit.location.province }
        if (canMoveViaConvoy(t.unit, t.destination.province, provinces.toSet)) {
          Option(Route(true))
        } else if (map.movableLocationsOf(t.unit) contains t.destination) {
          Option(Route(false))
        } else {
          None
        }
      case _ => None
    }
  }
}

File: src/variant/standard/rule/movement-order-group.ts
import { Result } from "./result"
import { MilitaryBranch } from "./military-branch"
import { StandardRuleUtils } from "./standard-rule-utils"
import { OrderType, Order, Convoy, Support, Move } from "./order"
import { DiplomacyMap } from "./types"
import { MovementOrderWithResult } from "./movement-order-with-result"

const { Fleet } = MilitaryBranch

export interface Route {
  viaConvoy: boolean
}
export class MovementOrderGroup<Power> {
  relatedOrders: Set<MovementOrderWithResult<Power>>
  constructor (
    public target: MovementOrderWithResult<Power>,
    relatedOrders: Set<MovementOrderWithResult<Power>> | Array<MovementOrderWithResult<Power>>
  ) {
    this.relatedOrders = new Set([...relatedOrders])
  }
  validSupports (): Set<Support<Power>> {
    return new Set(
      [...this.relatedOrders].filter(order => {
        return (order.order.tpe === OrderType.Support) &&
          (order.getResult() !== Result.Dislodged) && (order.getResult() !== Result.Cut) &&
          (order.getResult() !== Result.NoCorrespondingOrder)
      }).map(order => <Support<Power>>order.order)
    )
  }

  power (): number {
    return 1 + this.validSupports().size
  }

  route (map: DiplomacyMap<Power>): Route | null {
    if (this.target.order instanceof Move) {
      const units = [...this.relatedOrders].filter(order => {
        return (order.order.tpe === OrderType.Convoy) &&
          ((order.getResult() === Result.Failed) || (order.getResult() === Result.Success))
      }).map(order => order.order.unit)

      if (
        StandardRuleUtils.isMovableViaSea(
          map, this.target.order.unit.location.province, this.target.order.destination.province,
          new Set(units)
        )
      ) {
        return { viaConvoy: true }
      } else if (
        map.movableLocationsOf(this.target.order.unit.location, this.target.order.unit.militaryBranch).has(this.target.order.destination)
      ) {
        return { viaConvoy: false }
      } else {
        return null
      }
    } else {
      return null
    }
  }
}
